File: src/main/java/com/vc/medium/L79.scala
package com.vc.medium

object L79 {
  def exist(board: Array[Array[Char]], word: String): Boolean = {
    val n = board.length
    if(n == 0) return false
    val m = board(0).length
    val dirs = Array((0, 1), (1, 0), (0, -1), (-1, 0))
    import scala.collection.mutable
    val visited = new mutable.HashSet[(Int, Int)]()
    def solve(x: Int, y: Int, index: Int): Boolean = {
      if(index == word.length) return true
      if(!visited.contains((x, y))) {
        if(index + 1 == word.length && word(index) == board(x)(y)) return true
        visited += ((x, y))
        if(word(index) == board(x)(y)) {
          dirs.foreach(dir => {
            val xNew = x + dir._1
            val yNew = y + dir._2
            if(xNew >= 0 && xNew < n && yNew >= 0 && yNew < m) {
              if(solve(xNew, yNew, index + 1)) return true
            }
          })
        }
        visited -= ((x, y))
      }
      return false
    }
    board.indices.foreach(i => {
      board(i).indices.foreach(j => {
        if(solve(i, j, 0)) return true
      })
    })
    false
  }
}

File: src/main/java/com/vc/medium/L79.java
package com.vc.medium;

class L79 {

    private char[][] board;
    private String word;
    private boolean[][] visited;
    public boolean exist(char[][] board, String word) {
        /**
             A B C E
             S F E S
             A D E E

             ABCEFSADEESE
         */
        if(board == null || board.length == 0) return false;

        int n = board.length;
        int m = board[0].length;

        this.board = board;
        this.word = word;
        this.visited = new boolean[n][m];

        for(int i = 0; i < n; i++) {
            for(int j = 0; j < m; j++) {
                if(board[i][j] == word.charAt(0) && helper(1, i, j)) return true;
            }
        }
        return false;
    }

    private int[][] dirs = {{0, 1},{1, 0},{0, -1},{-1, 0}};
    private boolean helper(int index, int row, int col) {
        if(visited[row][col]) return false;

        if(index == word.length()) return true;

        visited[row][col] = true;
        for(int[] dir: dirs) {
            int rowNew = row + dir[0];
            int colNew = col + dir[1];
            if(rowNew >= 0 && rowNew < board.length && colNew >= 0
                    && colNew < board[0].length && board[rowNew][colNew] == word.charAt(index)) {
                if(helper(index + 1, rowNew, colNew)) return true;
            }
        }
        visited[row][col] = false;
        return false;
    }
}
